String formatting support

spernsteiner opened the pull request:
This branch adds support for string formatting, which we will eventually use for custom counterexample printing.  Along the way, it also:

 * Changes the representation of `&str` from `VectorType (BVType 32)`
   (Haskell-style strings) to `MirImmSlice (BVType 8)`.  This makes `&str` match the representation of `&[u8]`, as it does in normal Rust.  This also lets us implement a variety of `String` and `str` methods that were previously unsupported.

 * Fixes translation of function-pointer calls and some types of casts.

 * Exposes Crucible's `AnyType` to Rust, as `crucible::any::Any`.  We use this
   to reimplement the type erasure performed in the `core::fmt` module, which originally used `mem::transmute` in a way we don't support.

With this branch applied, we can now handle most uses of the `format!` and `format_args!` macros, at least in cases where the arguments are concrete.

atomb commented on src/Mir/TransCustom.hs:
```diff
@@ -325,6 +326,36 @@ vector_copy_from_slice = ( ["crucible","vector","{{impl}}", "copy_from_slice"],
+
+
+
 -----------------------------------------------------------------------------------------------------
```
It would be great if we had a more concise and readable way of writing stuff like this, though it might be that it would take more effort to develop that than it'd be worth.

atomb commented:
It would be good if we could fix the Travis build scripts to get testing feedback in PRs. I'm not sure offhand how difficult that would be.

spernsteiner commented:
Opened #11 for the boilerplate and #12 for Travis.  I'll merge this and then look at fixing those.

>One thing I wonder with all these changes to the standard libraries: how easily will we be able to adapt to newer versions of them? Is there anything we should do to more easily accommodate that?

I think we're in reasonably good shape, though we haven't actually done a Rust upgrade since replacing `mir-lib` (which was more-or-less written from scratch) with the edited upstream libraries we have now.  The parts we've modified are generally quite low-level, and I don't expect them to change much.  Most new additions to the library should build upon the functions we've already modified, rather than implementing similarly low-level functionality from scratch.  But we won't know for sure until we actually try the upgrade.
